ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java

```diff
@@ -0,0 +1,100 @@
+    final Query statement = (Query) KsqlParserTestUtil.buildSingleAst(query, metaStore).getStatement();
+    final Expression predicate = statement.getWhere().get();
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("(ORDERS.ROWTIME > STRINGTOTIMESTAMP('2017-01-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS'))"));
+  }
+
+  @Test
+  public void shouldHandleInexactTimestamp() {
+    final String query = "SELECT * FROM orders where ROWTIME = '2017-01-01';";
+    final Query statement = (Query) KsqlParserTestUtil.buildSingleAst(query, metaStore).getStatement();
+    final Expression predicate = statement.getWhere().get();
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("(ORDERS.ROWTIME = STRINGTOTIMESTAMP('2017-01-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS'))"));
+  }
+
+  @Test
+  public void shouldHandleBetweenExpression() {
+    final String query = "SELECT * FROM orders where ROWTIME BETWEEN '2017-01-01' AND '2017-02-01';";
+    final Query statement = (Query) KsqlParserTestUtil.buildSingleAst(query, metaStore).getStatement();
+    final Expression predicate = statement.getWhere().get();
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("(ORDERS.ROWTIME BETWEEN"
+        + " STRINGTOTIMESTAMP('2017-01-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS') AND"
+        + " STRINGTOTIMESTAMP('2017-02-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS'))"));
+  }
+
+  @Test
+  public void shouldNotProcessStringsInFunctions() {
+    final String query = "SELECT * FROM orders where ROWTIME = foo('bar');";
+    final Query statement = (Query) KsqlParserTestUtil.buildSingleAst(query, metaStore).getStatement();
+    final Expression predicate = statement.getWhere().get();
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("(ORDERS.ROWTIME = FOO('bar'))"));
+  }
+
+  @Test
+  public void shouldIgnoreNonRowtimeStrings() {
+    final String query = "SELECT * FROM orders where ROWTIME > '2017-01-01' AND ROWKEY = '2017-01-01';";
+    final Query statement = (Query) KsqlParserTestUtil.buildSingleAst(query, metaStore).getStatement();
+    final Expression predicate = statement.getWhere().get();
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("((ORDERS.ROWTIME > STRINGTOTIMESTAMP('2017-01-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS')) AND (ORDERS.ROWKEY = '2017-01-01'))"));
+  }
+
+  @Test
+  public void shouldHandleNestedExpressions() {
+    final String simpleQuery = "SELECT * FROM orders where FOO(ROWTIME + 6000) > '2017-01-01 00:00:00.000' + 35;";
```

Comment:
Isn't this a bit dangerous? We're assuming FOO is returning a `long` and that represents a datetime, but this might not be the case, e.g. `SELECT * FROM orders where IS_LEAP_YEAR(ROWTIME + 6000) > '2017-01-01 00:00:00.000' + 35;` wouldn't make sense!

I don't think we can make any such assumption on the return value from a function.

Reply:
My thoughts were that this is purely a query rewriter, so those kinds of cases would get caught later on. So for example, `IS_LEAP_YEAR(ROWTIME + 6000) > 35` passes this step, but would still eventually get caught during evaluation. I did change the overall logic of the rewriter as per @hjafarpour's comment though.

Reply:
They may get caught later on, but the UX for the user is not great.  There's going to get some message that `Foo can not be compared to Long` or something, which is very confusing for the user.